### ISSUE TITLE
Feature request: Issue a warning when @MinLen() is used on primitive types.

### ISSUE THREAD

**Issue by NhatDinh** [typetools/checker-framework#2075]
As discussed in the closed [Issue #1905](https://github.com/typetools/checker-framework/issues/1905), the annotation `MinLen()` of Index Checker should only be applied to sequence data structures( arrays, strings).
Consider:
```java
import org.checkerframework.common.value.qual.MinLen;

public class IssueWarningMinLenOnPrimitiveTypes {
    void MinLen(@MinLen(1) int[] array, @MinLen(1) String str){
        int val_int = array[0];//line 5
        @MinLen(1) char[] char_arr = str.toCharArray();//line 6
        char val_char = char_arr[0];
    }
}
```
To run this code:

```
javac -processor org.checkerframework.checker.index.IndexChecker IssueWarningMinLenOnPrimitiveTypes.java
```
Warnings issued:
```java
line 5: [array.access.unsafe.high.constant] Potentially unsafe array access: the constant index 0 could be larger than the array's bound
  found   : @ArrayLenRange(from=1, to=2147483647) int @UnknownVal []
  required: @MinLen("1") -- an array guaranteed to have at least 1 elements
line 6: [assignment.type.incompatible] incompatible types in assignment.
  found   : @UnknownVal char @ArrayLenRange(from=1, to=2147483647) []
  required: @ArrayLenRange(from=1, to=2147483647) char @UnknownVal []
```

**Comment by kelloggm** [comment]
What if we add a special case to these errors that checks if the component type matches the required type and issues a warning suggesting that the annotations are switched?

Alternatively, we could just add logic to the visitor to prevent non-sensical types from being applied.

**Comment by mernst** [comment]
I think you should be able to use [`@RelevantTypes`](https://checkerframework.org/api/org/checkerframework/framework/qual/RelevantJavaTypes.html).
The Checker Framework should issue an error when an annotation is written on a type other than a relevant type.  If it does not, that is a bug.

> Alternatively, we could just add logic to the visitor to prevent non-sensical types from being applied.

This is undesirable.  It would be confusing for users if an annotation were silently ignored.

**Comment by kelloggm** [comment]
Perhaps I was unclear. By "prevent non-sensical types from being applied" I meant issue an error at the location that the annotation is written.

I do not think that `@RelevantTypes` will work without modification. `@RelevantTypes` applies to a type checker (and therefore all its qualifiers); we would like to enforce different restrictions on `@MinLen` than on the other qualifiers in the Value Checker.

**Comment by mernst** [comment]
You are right.

Here are two possibilities:

1. Create and enforce a `@TargetTypeKinds` annotation, which is analogous to
[`@TargetLocations`](https://checkerframework.org/api/org/checkerframework/framework/qual/TargetLocations.html).
BaseTypeVisitor can warn when a use is not consistent with its `@TargetTypeKinds` meta-annotation.
`@MinLen` would be declared as:
```
@TargetTypeKinds(TypeKind.ARRAY)
public @interface MinLen { ... }
```

2. Permit `@RelevantJavaTypes` to appear on a type qualifier.
This is richer than the proposed ```@TargetTypeKinds``` cannot express that an annotation can appear on both `int` and `Integer`.

I lean toward the latter.
